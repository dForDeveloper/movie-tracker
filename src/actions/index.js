export const addMovie = (movie) => ({
  type: 'ADD_MOVIE',
  movie
});

export const setUser = (user) => ({
  type: 'SET_USER',
  user
});

export const logoutUser = () => ({
  type: 'LOGOUT_USER'
});

export const getFavorites = (favorites) => ({
  type: 'GET_FAVORITES',
  favorites
});

export const addFavorite = (movie) => ({
  type: 'ADD_FAVORITE',
  movie
});