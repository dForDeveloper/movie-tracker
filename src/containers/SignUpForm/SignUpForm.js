import React, { Component } from 'react';
import { createUser } from '../../utils/queries';
import { Redirect } from 'react-router-dom';

class SignUpForm extends Component {
  constructor() {
    super();
    this.state = {
      name: '',
      email: '',
      passwordOriginal: '',
      passwordConfirmed: '',
      status: ''
    };
  }

  handleChange = (event) => {
    const { id, value } = event.target;
    this.setState({ [id]: value });
  }

  handleSubmit = async (event) => {
    event.preventDefault();
    const { name, email, passwordOriginal } = this.state;
    if (this.checkMatchingPassword() && this.checkEmailRegex()) {
      try {
        const response = await createUser(name, email, passwordOriginal);
        this.setState({ status: response.status });
      } catch (error) {
        this.setState({ status: 'error' });
      }
    }
  }

  checkEmailRegex = () => {
    const emailRegex = /[a-z]\w+@[a-z]\w+\.[a-z]\w+/;
    return emailRegex.test(this.state.email);
  }

  checkMatchingPassword = () => {
    const { passwordOriginal, passwordConfirmed } = this.state;
    return passwordOriginal === passwordConfirmed;  
  }

  getNameAndEmailInputFields = () => {
    const { name, email } = this.state;
    const inputStyle = this.checkEmailRegex() ? null : 'input--red';
    return (
      <div>
        <label htmlFor="name">Name:</label>
        <input id="name" value={name} onChange={this.handleChange} required />
        <label htmlFor="email">Email</label>
        <input
          id="email"
          value={email}
          onChange={this.handleChange}
          required
          className={inputStyle}
        />
        {!this.checkEmailRegex() && <p>Enter a valid email</p>}
      </div>
    )
  }

  getPasswordInputFields = () => {
    const { passwordOriginal, passwordConfirmed } = this.state;
    const inputStyle = this.checkMatchingPassword() ? null : 'input--red';
    return (
      <div className="div--password">
        <label htmlFor="passwordOriginal">Password</label>
        <input 
          id="passwordOriginal"
          value={passwordOriginal}
          onChange={this.handleChange}
          type="password" 
          required
        />
        <label htmlFor="passwordConfirmed">Confirm Password</label>
        <input 
          id="passwordConfirmed"
          value={passwordConfirmed}
          onChange={this.handleChange}
          type="password" 
          required
          className={inputStyle}
        />
        {!this.checkMatchingPassword() && <p>Passwords do not match</p>}
      </div>
    )
  }

  render() {
    const { name, email, status } = this.state;
    return(
      <form onSubmit={this.handleSubmit} className="form">
        <h2>Sign Up</h2>
        {this.getNameAndEmailInputFields()}
        {this.getPasswordInputFields()}
        <input type="submit" value="Submit" />
        {status === 'error' && <p>Email already registered to an account</p>}
        {status === 'success' && <Redirect to='/login' />}
      </form>
    );
  }
}

export default SignUpForm;