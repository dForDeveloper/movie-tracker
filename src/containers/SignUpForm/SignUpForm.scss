.SignUpForm {
  display: grid;
  justify-content: center;
}

input {
  display: block;
  margin-bottom: 10px;
}

.input--red {
  outline-color: red;
  border: 2px solid red;
}

.div--password {
  margin-top: 30px;
}