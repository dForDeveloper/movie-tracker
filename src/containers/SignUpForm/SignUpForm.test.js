import React from 'react';
import { shallow } from 'enzyme';
import { SignUpForm, mapDispatchToProps } from './SignUpForm';
import { setUser } from '../../actions';

const mockProps = {
  history: {},
  location: {},
  match: {},
  setUser: jest.fn()
}

describe('SignUpForm', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <SignUpForm {...mockProps} />
    );
  });
  
  describe('SignUpForm container', () => {
    it('should properly render the component elements', () => {
      expect(wrapper).toMatchSnapshot();
    });
  
    it('should have the proper default state', () => {
      expect(wrapper.state()).toEqual({
        name: '',
        email: '',
        passwordOriginal: '',
        passwordConfirmed: '',
        status: ''
      })
    })
  
    it('should set state based on an input field when text is typed', () => {
      const event = {target: {id: 'name', value: 'shannon'}};
      wrapper.find('#name').simulate('change', event);
      expect(wrapper.state('name')).toEqual('shannon');
    });
    
    it('should return true when passwords match', () => {
      const mockPasswordMatch = wrapper.instance().checkMatchingPassword();
      expect(mockPasswordMatch).toEqual(true);
    });

    it('should return false when passwords do not match', () => {
      const event = {target: {id: 'passwordOriginal', value: 'password123'}};
      wrapper.find('#passwordOriginal').simulate('change', event);
      const mockPasswordMatch = wrapper.instance().checkMatchingPassword();
      expect(mockPasswordMatch).toEqual(false);
    });
    
    it('should return true when a user enters a valid email', () => {
      const mockEmail = 'valid@test.io';
      const result = wrapper.instance().checkEmailRegex(mockEmail);
      expect(result).toEqual(true);
    });

    it('should return false when a user enters a invalid email', () => {
      const mockEmail = 'invalid-email';
      const result = wrapper.instance().checkEmailRegex(mockEmail);
      expect(result).toEqual(false);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch when setUser is called', () => {
      const dispatchMock = jest.fn();
      const expected = setUser({ id: 1, name: 'Jeo' });
      const result = mapDispatchToProps(dispatchMock);
      result.setUser({ id: 1, name: 'Jeo' });
      expect(dispatchMock).toHaveBeenCalledWith(expected);
    });
  });
});