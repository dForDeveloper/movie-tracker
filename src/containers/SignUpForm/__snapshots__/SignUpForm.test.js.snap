// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm SignUpForm container should properly render the component elements 1`] = `
<form
  className="SignUpForm"
  onSubmit={[Function]}
>
  <h2
    className="signup-title"
  >
    Sign Up
  </h2>
  <div
    className="div--email"
  >
    <label
      htmlFor="name"
    >
      Name:
    </label>
    <input
      id="name"
      onChange={[Function]}
      required={true}
      value=""
    />
    <label
      htmlFor="email"
    >
      Email
    </label>
    <input
      className="input--red"
      id="email"
      onChange={[Function]}
      required={true}
      value=""
    />
    <p
      className="notice"
    >
      Please enter a valid email.
    </p>
  </div>
  <div
    className="div--password"
  >
    <label
      htmlFor="passwordOriginal"
    >
      Password
    </label>
    <input
      id="passwordOriginal"
      onChange={[Function]}
      required={true}
      type="password"
      value=""
    />
    <label
      htmlFor="passwordConfirmed"
    >
      Confirm Password
    </label>
    <input
      className={null}
      id="passwordConfirmed"
      onChange={[Function]}
      required={true}
      type="password"
      value=""
    />
  </div>
  <input
    className="create-button"
    type="submit"
    value="Submit"
  />
</form>
`;
